### PR TITLE
adapter: promote mz_cluster_replica_frontiers

### LINKED ISSUE #29233: catalog: promote mz_internal.mz_cluster_replica_frontiers builtin source to mz_catalog
The `mz_internal.mz_cluster_replica_frontiers` builtin source is uniquely used to power when the 0DT cutover happens. As such, just promoting it the usual way by switching it to the `mz_catalog` schema causes issues in the upgrade process, which were caught by the legacy upgrade tests.

See more discussion and failing test here https://materializeinc.slack.com/archives/C02FWJ94HME/p1724705003113639

As such, this source should be promoted to `mz_catalog` as part of the high availability during version upgrade project.

Note: we also discovered a complication that builtin sources are truncated when changed, and the promotion to `mz_catalog` is a change. Perhaps special handling of builtin sources is useful pre-work, if we don't want truncation on promotion.

### PR DESCRIPTION
**Pull request by benesch**
This was on our list of relations to promote in August, but it requires a bit of special care due to the way the relation is used by the 0dt machinery.

Fix (eventually) #29233 (leaving issue open to track the removal of the migration).

### Motivation

  * This PR adds a known desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/adapter/src/catalog/migrate.rs [hunk 1]
```diff
@@ -185,9 +185,28 @@ pub(crate) async fn migrate(
...
+    // over to the new version.
+    //
+    // TODO(benesch): remove this for the next release (27 September 2024).
+    let mut mappings = tx.get_system_object_mappings().collect::<Vec<_>>();
+    if let Some(object) = mappings.iter_mut().find(|o| o.description == SystemObjectDescription {
+        schema_name: "mz_internal".into(),
+        object_name: "mz_cluster_replica_frontiers".into(),
+        object_type: mz_sql::catalog::CatalogItemType::Source,
+    }) {
+        object.description.schema_name = "mz_catalog".into();
```
Comment (jkosh44) [inline src/adapter/src/catalog/migrate.rs:206]: Not a big deal, but these all have constants laying around, if they're accessible.

- `mz_repr::namespaces::MZ_INTERNAL_SCHEMA`
- `mz_repr::namespaces::MZ_CATALOG_SCHEMA`
- `mz_catalog::builtin::MZ_CLUSTER_REPLICA_FRONTIERS.name`

**Review by jkosh44 — APPROVED** [review]
LGTM

**Comment by benesch** [comment]
I fixed the test failures and triggered a new nightly run: https://buildkite.com/materialize/nightly/builds/9624

**Comment by benesch** [comment]
I fixed the nightly failures and triggered _another_ new nightly run: https://buildkite.com/materialize/nightly/builds/9626

**Pull request merged**